src/navigation_plugin.cpp

```diff
@@ -68,84 +64,73 @@ void constructConstraint(std::vector<geo::Vector3>& chull, std::stringstream& co
 
 // ----------------------------------------------------------------------------------------------------
 
+
 //
 /**
  * @brief constructShapeConstraint
- * @param r r must be currently reading an array of shapes
+ * @param shape shape of the volume in entity frame
  * @param entity_pose pose of the entity
  * @return constraint string
  */
-std::string constructShapeConstraint(tue::config::Reader& r, const geo::Pose3D& entity_pose)
+std::string constructShapeConstraint(geo::ShapeConstPtr& shape, const geo::Pose3D& entity_pose)
 {
     std::stringstream shape_constraint;
 
-    bool first_sub_shape = true;
-    while(r.nextArrayItem())
-    {
-        std::stringstream sub_shape_constraint;
+    geo::Shape shape_tr;
+    shape_tr.setMesh(shape->getMesh().getTransformed(entity_pose)); // get shape in map frame.
+    std::vector<geo::Vector3> points = shape_tr.getMesh().getPoints();
 
-        if (r.readGroup("box", tue::config::OPTIONAL))
-        {
-            std::vector<geo::Vector3> chull; // In MAP frame
+    // Calculate cluster convex hull
```

Comment:
Looks like a generic function, so it should not be coded here. As you throw the result in a hull, perhaps extend that to do such computations if necessary? (assuming you'll have more sources of hulls without these properties already computed.)